### PR TITLE
stopcount

### PR DESCRIPTION
**Pull request by silabs-krdosvik**
add dcsr stopcount test
Add to full regression
Add to vplan.

### CODE REVIEW

File: docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json
Comment (silabs-robin) [inline docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json:1]: Thanks for updating the vplan too! I myself am very liable to forget that, but it should be second nature, really.

File: docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json [hunk 1]
```diff
@@ -376,19 +376,19 @@
         "Feature": "Debug Mode",
         "Sub Feature": "",
         "Feature Description": "",
-        "Verification Goal": "hpmcounter[31-3] continues during debug",
+        "Verification Goal": "hpmcounter[31-3] continues during debug if stopcount=0, and stop if stopcount=1",
         "Pass/Fail Criteria": "Self Checking Test",
         "Test Type": "Directed Self-Checking",
         "Coverage  Method": "Testcase",
-        "Link to Coverage": "",
+        "Link to Coverage": "DT: mhpmcounter3_dcsr_stopcount.c, TODO: test hpmcounter[31-4]",
```
Comment (silabs-robin) [inline docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json:383]: What is this TODO?

**Review by silabs-robin — COMMENTED** [review]
I've reviewed the non-C files.
Doing the C files next.

File: docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json
Comment (silabs-krdosvik) [reply docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json:1]: Thanks! Yea, I guess we have potential when it comes to that.

File: docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json (hunk 1 above)
Comment (silabs-krdosvik) [reply docs/VerifPlans/Simulation/privileged_spec/CV32E40X_Counters.json:383]: Aaa, yea, I have made the test mhpmcounter 3 specific, so I have not tested that stopcount works on counter 4 to 31. The todo is there to indicate that.

File: tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  __asm__ volatile(R"(
+    mv t0, x0
+    csrw 0xB03, t0
+  )"::: "t0");
+
+  //Read the counter's value
+  __asm__ volatile("csrr %[counter], 0xB03" : [counter]"=r"(counter));
+
+  if (counter != 0){
+    return 1;
```
Comment (silabs-robin) [inline tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c:208]: Is this handled?

File: tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+ */
+int counter3_minstret_tests (int g_is_dcsr_stopcount)
+  {
+  int is_failure = 0;
+  int counter = 0;
+  int cnt_retired_instr = (1 << MHPMEVENT_INSTR_BIT_POS);
+  int expected_minstret = 3 + 1; //3 nop instructions and 1 instruction to deactivate counter
+
+  //Make the counter count number of retired instructions
+  __asm__ volatile("csrw 0x323, %0 " :: "r"(cnt_retired_instr));
```
Comment (silabs-robin) [inline tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c:233]: Is this the hex for `mhpmevent3`? It should be possible to write the csr name where the hex is.

File: tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  //Activate the counter, run some instructions, deactivate the counter
+  __asm__ volatile(R"(
+    mv t0, x0
+    not t1, t0
+    csrw mcountinhibit, t0
+    nop
+    nop
+    nop
+    csrw mcountinhibit, t1
+  )"::: "t0", "t1");
```
Comment (silabs-robin) [inline tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c:244]: I'm not 100% solid on the use of clobbers, but this does indeed match how I believe them to work.

File: tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c (hunk 2 above)
Comment (silabs-krdosvik) [reply tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c:208]: A good rhetorical question. FIXED

File: tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c (hunk 3 above)
Comment (silabs-krdosvik) [reply tests/programs/custom/mhpmcounter3_dcsr_stopcount/mhpmcounter3_dcsr_stopcount.c]: True, agreed that it is better like that! FIXED

**Review by silabs-robin — APPROVED** [review]
Approving based on:
All the details I've spot-checked seems in order.
All feedback resolved.
Presuming the test passes.

**Pull request merged**